Sources/Afluent/SequenceOperators/TimerSequence.swift

```diff
@@ -0,0 +1,84 @@
+//
+//  TimerSequence.swift
+//  Afluent
+//
+//  Created by Annalise Mariottini on 11/9/24.
+//
+
+import Foundation
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, visionOS 1.0, *)
+extension AsyncSequences {
+    /// A sequence that repeatedly emits a time duration offsetn on a given interval.
+    public struct TimerSequence<C: Clock>: AsyncSequence, Sendable {
+        public typealias Element = C.Instant
+
+        init(interval: C.Duration, tolerance: C.Duration?, clock: C) {
+            self.interval = interval
+            self.tolerance = tolerance
+            self.clock = clock
+        }
+
+        private let interval: C.Duration
+        private let tolerance: C.Duration?
+        private let clock: C
+
+        public struct AsyncIterator: AsyncIteratorProtocol {
+            private var cancellables: Set<AnyCancellable> = []
+
+            init(interval: C.Duration, tolerance: C.Duration?, clock: C) {
+                let (stream, continuation) = AsyncStream<C.Instant>.makeStream()
+                self.iterator = stream.makeAsyncIterator()
+
+                DeferredTask {
```

Comment:
Can we avoid the extra task here? I think this should all be possible in the next function. 

I can pull up an example but there's a whole thing over on swift-async-algorithms about avoiding any new tasks

Reply:
Question: do you want to avoid this entirely since it already exists in swift-async-algorithms?

Reply:
Unless we can't? Maybe the task is necessary because of when next is called. I am cool leaving this to emulate what Apple did over there

Reply:
Yeah, so [their implementation](https://github.com/apple/swift-async-algorithms/blob/4c3ea81f81f0a25d0470188459c6d4bf20cf2f97/Sources/AsyncAlgorithms/AsyncTimerSequence.swift#L14) is honestly a bit strange to me, since it relies on the call to `next()` to generate the next element. OMHO, a timer sequence should be emitting elements whether or not there's a call to `next()`, since it's based on the advancement of the clock.

E.g., in the following scenario, what should occur?

```swift
let testClock = TestClock()
let seq = TimerPublisher.publish(every: .seconds(1), clock: testClock)

let it = seq.makeAsyncIterator()

testClock.advance(by: .seconds(5))

// should this print offset 1 second? or offset 5 seconds?
print(try await it.next())
```

I _think_ the swift-async-algorithms version prints 5? (Will look more at their implementation tomorrow, maybe I'm wrong.)

Reply:
Maybe it does make to let elements to "pass through" if there's no listener. That would mimic the publisher implementation.

Reply:
_On the other hand_, we would lose the ability to, for example, `print()` the elements in the sequence, even when not actively being iterated on.

Reply:
So you aren't actually emitting without a subscription with the current implementation. The stream has a buffer you are appending to

Reply:
I think the example of not being able to print is okay. It follows Combine's notion of not doing anything without a subscription

Reply:
Sorry, should have been more precise with my language: not "emitting elements" but rather "producing elements for future consumption" (as you said, stored in a buffer).

My understanding of this appears to be misguided. Looking at Combine a bit more closely, I've noticed a few things about the behavior:

1. A subscription request is what starts the timer to start producing elements
2. Breaks in subscription result in breaks in the emitted elements
3. The exact timing of the emitted elements depends directly upon the time of subscription

Here's an example to show what I mean:
```swift
let formatter = DateFormatter()
formatter.dateFormat = "H:mm:ss.SSSS"

let publisher = Timer.publish(every: 1.0, on: .main, in: .common)
    .autoconnect()

print("DATE \(formatter.string(from: Date()))")

try await Task.sleep(for: .seconds(2.5))

publisher.sink { date in
    print("SINK DATE 1 \(formatter.string(from: date))")
}.store(in: &cancellables)

try await Task.sleep(for: .seconds(2.5))
cancellables.removeAll()

publisher.sink { date in
    print("SINK DATE 2 \(formatter.string(from: date))")
}.store(in: &cancellables)

try await Task.sleep(for: .seconds(2.5))
cancellables.removeAll()
```

One run of the above produces:
```
DATE 15:44:42.3540
SINK DATE 1 15:44:45.9300
SINK DATE 1 15:44:46.9260
SINK DATE 2 15:44:48.4660
SINK DATE 2 15:44:49.4630
```

Notice that the milliseconds on the emitted event do not align, since the subscriptions are created off cadence from each other.

Interestingly, if we keep the omit the first `cancellables.removeAll()` line, keeping the first subscription alive when the second is made, they are aligned:
```
DATE 15:47:29.9600
SINK DATE 1 15:47:33.4830
SINK DATE 1 15:47:34.4790
SINK DATE 1 15:47:35.4790
SINK DATE 2 15:47:35.4790
SINK DATE 1 15:47:36.4790
SINK DATE 2 15:47:36.4790
SINK DATE 1 15:47:37.4790
SINK DATE 2 15:47:37.4790
```

Reply:
Given the above behavior, I believe the swift-async-algorithms implementation results in the correct behavior. Going to try adjusting now, and writing a test to prove that we replicate the above behavior expectation.

Reply:
Just pushed a change. @Tyler-Keith-Thompson I'd appreciate a re-review when you can.

Reply:
This looks like top notch work. Thank you!